### ISSUE TITLE
WINDOWID environment variable is set to 0 instead of the actual Window id

### ISSUE THREAD

**Issue by Lyle-Tafoya** [lxqt/qterminal#659]
##### Expected Behavior
`echo $WINDOWID` should ideally produce the terminal's X11 Window id.

##### Current Behavior
`echo $WINDOWID` produces an output of `0` (which actually means it's being set by qterminal already. Just not set to a meaningful value)

##### Context
I'm trying to develop an input system for my game engine which would be suitable for real-time ASCII games that run in a terminal emulator. It does not seem to be possible to reliably get key release events from most/any terminal emulators without getting xlib or xcb involved. As part of this, I need to get the Window id before I can subscribe to any events associated with that Window. By far the easiest way to do this is to read the contents of the WINDOWID environment variable which is set by many terminal emulators currently (ie. Konsole, xterm, Allacrity, kitty, etc...). The alternative approach is to attempt to lookup the Window id from the pid, but that approach is not ideal because it has to make some assumptions about which pid is the correct one. Those assumptions will not hold up in every circumstance that the application might be run. This is why having the WINDOWID variable set correctly would help me.

##### System Information

* Distribution & Version: Arch Linux
* Kernel: 5.5.2-arch1-1
* Qt Version: 5.14.1
* qtermwidget Version: 0.14.1
* lxqt-build-tools Version: ?
* Package version: ?

**Comment by agaida** [comment]
erm - no. and if n terminals are open which windowi do you expect?

**Comment by tsujan** [comment]
@agaida At first, I had the same thought but then, I thought he might mean the terminal window itself. Even if he didn't mean that, XTerm and Konsole give an ID; QTerminal doesn't. So, I think this is a valid bug report -- unless I'm missing something basic.

**Comment by agaida** [comment]
ok, i see no substantial value in it - but it might be possible to set the WINDOWID, if qterminal is started from a terminal - might be a nice gimmick. And now to something completely different - the Monty Pythons Crew would like to suggest, that you start the first terminal from a terminal with qtermina & and the second with qterminal & - just have fun with the WINDOWID :smile:

**Comment by tsujan** [comment]
> you start the first terminal from a terminal with qterminal &

Yes, that works.

> i see no substantial value in it 

Me neither. But I think there is (or was) another EV that QTerminal didn't show but Xterm did. Sadly, I don't remember what it was -- will tell you if I remember it. Anyhow, that was the reason why the report caught my attention.

**Comment by tsujan** [comment]
UPDATE: It seems that there's no EV missing from QTerminal now.

**Comment by agaida** [comment]
hmm - the following answer is limited by my limited knowledge in english - the only thing that would make sense would be:
* one open qterminal and the EV WINDOWID is set to the just opened terminal
* next terminal opened - next WINDOWID - i've thought of an external thing, my bad, my mistake. so every terminal deliver it's own window id - would make sense maybe

**Comment by tsujan** [comment]
> would make sense maybe

Yes, it does. Every window, under X11, has an ID.

**Comment by tsujan** [comment]
This can be closed. Detailed explanation:

If you take a look at https://github.com/lxqt/qtermwidget/blob/911ae03ccfd532f8fc242547c8859667f83bae41/lib/Session.cpp#L124, you see that, in `qtermwidget` → `Session.cpp` → `WId Session::windowId()`, we intentionally return 0 because:

 1. If we use `QWidget::winId()` for a view in that function (like Konsole does), a native handle will be forcefully created and that might break some GUIs that use `qtermwidget` (like `QQuickWidget`).

 2. If we safely use `QWidget::internalWinId()` or `QWidget::effectiveWinId()` for a view, we'll still get 0 (I tested it) because the window isn't created yet.
 
In short, unlike XTerm, Konsole and most terminal emulators, `qtermwidget` may be used anywhere and we don't want to break that. Otherwise, a valid window ID would be possible by:

```c++
WId Session::windowId() const
{
    if (!_views.isEmpty())
    {
        QWidget* w = _views.first();
        if (w != nullptr)
        {
            while (w->parentWidget() != nullptr)
                w = w->parentWidget();
            return w->winId(); // WARNING: This may create a native handle where it shouldn't!
        }
    }
    return 0;
}
```

**Comment by tsujan** [comment]
BTW, I encountered the dangers of `QWidget::winId()` in Kvantum a long time ago and wouldn't use it anywhere. The undocumented function `QWidget::internalWinId()` is always safe but it doesn't make a difference with `qtermwidget` (it does with Kvantum).
